### PR TITLE
Support annotationProcessor configuration.

### LINKED ISSUE #3026: Feature Request: Native support on annotationProcessor Gradle Task (2.2.0)
Hi Realm team -  If you can have this compatible with annotationProcessor as well aside from the apt plugin, it can reduce the dependency on additional 3rd party plugin for the annotation processing natively supported on Android N. Thanks.

Here's the setup to validate the support:
parent build,gradle should be using 
`classpath 'com.android.tools.build:gradle:2.2.0-alpha1'`

The following line is removed on the application build,gradle
`apply plugin: 'com.neenbedankt.android-apt'`

### PR DESCRIPTION
**Pull request by zaki50**
Supported `annotationProcessor` configuration provided by Android Gradle Plugin 2.2.0 or later.

Realm plugin adds its annotation processor to `annotationProcessor` configuration instead of `apt` configuration if `annotationProcessor` configuration is provided and `com.neenbedankt.android-apt` plugin is not applied.

In Kotlin project, `annotationProcessor` configuration is never used for now

The projects that satisfy all following conditions must apply `com.neenbedankt.android-apt` manually (or modify the configurations to `annotationProsessor` or its variants).

* Java project (not Kotlin project).
* Not applying `com.neenbedankt.android-apt` plugin manually.
* Using `apt` or its variants (`*Apt`) configurations.

fixes #3026 

In this PR, Realm gradle plugin chooses a configuration by following rules.

First, `com.neenbedankt.android-apt` plugin was already applied manually by the user, Realm plugin always uses `apt` configuration.

If `com.neenbedankt.android-apt` was not applied by the user, Realm plugin uses :

| project language |`preferAptOnKotlinProject` | `annotationProcessor` configuration defined |  configuration used by Realm Plugin
| ----- | -------- | ------ | ----
| Java |  ignored | true | `annotationProcessor`
| Java |  ignored | false | `apt` (with applying `android-apt`)
| Kotlin | true | true | `annotationProcessor`
| Kotlin | true | false | `apt` (with applying `android-apt`)
| Kotlin | false | ignored | `kapt`

**Currently we don't provide a way to set `preferAptOnKotlinProject` to `true`.**

### CODE REVIEW

**Comment by zaki50** [comment]
review this please @realm/java 

I locally tested it with Android Gradle Plugin 1.5.0, 2.1.0, 2.2.0, 2.2.2 (with/without apt-plugin).

File: gradle-plugin/src/main/groovy/io/realm/gradle/Realm.groovy [hunk 1]
```diff
@@ -44,27 +42,36 @@ class Realm implements Plugin<Project> {
         def syncEnabledDefault = false
         project.extensions.create('realm', RealmPluginExtension, project, syncEnabledDefault)
 
-        def usesKotlinPlugin = project.plugins.findPlugin('kotlin-android') != null
         def usesAptPlugin = project.plugins.findPlugin('com.neenbedankt.android-apt') != null
+        def usesKotlinPlugin = project.plugins.findPlugin('kotlin-android') != null
+        def hasAnnotationProcessorConfiguration = project.getConfigurations().findByName('annotationProcessor') != null
+        // TODO add a parameter in 'realm' block if this should be specified by users
+        def preferAptOnKotlinProject = false
```
Comment (cmelchior) [inline gradle-plugin/src/main/groovy/io/realm/gradle/Realm.groovy]: What happens if you use java model classes in a Kotlin project? If it doesn't work, I think we should just add this to the config block straight away.

Also, wouldn't it be necessary to specify if `apt` or `annotationProcessor` should be used?

File: gradle-plugin/src/main/groovy/io/realm/gradle/Realm.groovy [hunk 2]
```diff
@@ -44,27 +42,36 @@ class Realm implements Plugin<Project> {
...
 
         project.android.registerTransform(new RealmTransformer(project))
 
         project.repositories.add(project.getRepositories().jcenter())
         project.dependencies.add("compile", "io.realm:realm-annotations:${Version.VERSION}")
-        if (isKaptProject) {
-            project.dependencies.add("kapt", "io.realm:realm-annotations:${Version.VERSION}")
-            project.dependencies.add("kapt", "io.realm:realm-annotations-processor:${Version.VERSION}")
-        } else {
+        if (usesAptPlugin) {
```
Comment (cmelchior) [inline gradle-plugin/src/main/groovy/io/realm/gradle/Realm.groovy]: What if a Kotlin project uses `apt` for some java dependencies (e.g Butterknife), but `kapt` for Realm model classes?

File: gradle-plugin/src/main/groovy/io/realm/gradle/Realm.groovy [hunk 3]
```diff
@@ -44,27 +42,36 @@ class Realm implements Plugin<Project> {
...
             project.dependencies.add("apt", "io.realm:realm-annotations:${Version.VERSION}")
             project.dependencies.add("apt", "io.realm:realm-annotations-processor:${Version.VERSION}")
             project.dependencies.add("androidTestApt", "io.realm:realm-annotations:${Version.VERSION}")
             project.dependencies.add("androidTestApt", "io.realm:realm-annotations-processor:${Version.VERSION}")
+        } else if (usesKotlinPlugin && !preferAptOnKotlinProject) {
```
Comment (cmelchior) [inline gradle-plugin/src/main/groovy/io/realm/gradle/Realm.groovy]: I have a feeling this case should be the first in the if/else block since it is the most specific.

File: gradle-plugin/src/main/groovy/io/realm/gradle/Realm.groovy [hunk 4]
```diff
@@ -76,4 +83,19 @@ class Realm implements Plugin<Project> {
             return false
         }
     }
+
+    private static boolean shouldApplyAndroidAptPlugin(boolean usesAptPlugin, boolean usesKotlinPlugin,
+                                                       boolean hasAnnotationProcessorConfiguration,
+                                                       boolean preferAptOnKotlinProject) {
+        if (usesAptPlugin) {
+            // for any projects that use `android-apt` plugin
```
Comment (cmelchior) [inline gradle-plugin/src/main/groovy/io/realm/gradle/Realm.groovy]: I assume you mean `for any projects that uses android-apt plugin already. No need to apply it twice`. Just to clarify it? Because just reading it at a glance looks a bit weird.

File: gradle-plugin/src/main/groovy/io/realm/gradle/Realm.groovy (hunk 1 above)
Comment (zaki50) [inline gradle-plugin/src/main/groovy/io/realm/gradle/Realm.groovy]: In that case, this implementation (and of course previous implementation too) expects users manually apply `android-apt` plugin.

If we need to provide a way for users to choose `annotationProcessor` configuration in Kotlin project, exposing config parameter is mandatory.

File: gradle-plugin/src/main/groovy/io/realm/gradle/Realm.groovy (hunk 2 above)
Comment (zaki50) [inline gradle-plugin/src/main/groovy/io/realm/gradle/Realm.groovy]: I don't think that mixing `kapt` and `apt` works.
Is it possible?

Comment (cmelchior) [inline gradle-plugin/src/main/groovy/io/realm/gradle/Realm.groovy]: Perhaps not. I haven't tried it 😄

File: CHANGELOG.md [hunk 5]
```diff
@@ -1,3 +1,9 @@
+## 2.2.0
+
+### Enhancements
+
+* Added support for the `annotationProcessor` configuration provided by Android Gradle Plugin 2.2.0 or later. Realm plugin adds its annotation processor to the `annotationProcessor` configuration instead of `apt` configuration if it is available and the `com.neenbedankt.android-apt` plugin is not used. In Kotlin projects, `kapt` is used instead of the `annotationProcessor` configurationi (#3026).
```
Comment (dalinaum) [inline CHANGELOG.md]: `configurationi` -> `configuration`

Comment (zaki50) [inline CHANGELOG.md]: fixed!

**Comment by Gounlaf** [comment]
(Comment duplicated from issue #3026; don't know what the best place for that; sorry =/)

Hi @zaki50,

**TL;DR**: is plugin compatible with snapshot build ?

So, some explanations: 

I've installed the latest snapshot on new Android project (just an empty one), with builds tools 2.2.2.

> ~/RealmTest/app/build.gradle
> Error:Failed to resolve: annotationProcessor

The project level build.gradle:

```
// Top-level build file where you can add configuration options common to all sub-projects/modules.

buildscript {
    repositories {
        jcenter()
        maven {
            url 'http://oss.jfrog.org/artifactory/oss-snapshot-local'
        }
    }
    dependencies {
        classpath 'com.android.tools.build:gradle:2.2.2'

        // NOTE: Do not place your application dependencies here; they belong
        // in the individual module build.gradle files

        classpath "io.realm:realm-gradle-plugin:2.2.0-SNAPSHOT"
    }
}

allprojects {
    repositories {
        jcenter()
        maven {
            url 'http://oss.jfrog.org/artifactory/oss-snapshot-local'
        }
    }
}

task clean(type: Delete) {
    delete rootProject.buildDir
}
```

The application level build.gradle

```
apply plugin: 'com.android.application'
apply plugin: 'realm-android'

android {
    compileSdkVersion 25
    buildToolsVersion "25.0.0"
    defaultConfig {
        applicationId "testing.realm.realmtest"
        minSdkVersion 16
        targetSdkVersion 25
        versionCode 1
        versionName "1.0"
        testInstrumentationRunner "android.support.test.runner.AndroidJUnitRunner"
    }
    buildTypes {
        release {
            minifyEnabled false
            proguardFiles getDefaultProguardFile('proguard-android.txt'), 'proguard-rules.pro'
        }
    }
}

dependencies {
    compile fileTree(dir: 'libs', include: ['*.jar'])
    androidTestCompile('com.android.support.test.espresso:espresso-core:2.2.2', {
        exclude group: 'com.android.support', module: 'support-annotations'
    })
    compile 'com.android.support:appcompat-v7:25.0.0'
    testCompile 'junit:junit:4.12'
}
```

Looking into the Gradle plugin, I understand it put dependencies automatically

So, in application level build.gradle, desactivating plugin and putting dependencies myself :

```
apply plugin: 'com.android.application'
//apply plugin: 'realm-android'

dependencies {
    [...]
    testCompile 'junit:junit:4.12'

    compile "io.realm:realm-annotations:2.2.0-SNAPSHOT"
    annotationProcessor "io.realm:realm-annotations:2.2.0-SNAPSHOT"
    annotationProcessor "io.realm:realm-annotations-processor:2.2.0-SNAPSHOT"
}
```

**Without** ```annotationProcessor "io.realm:realm-annotations-processor:2.2.0-SNAPSHOT"``` gradle sync correctly, but application compilation failed ; **with** it, sync failed (and so compilation too).

Thanks.
